### PR TITLE
Add Verifiable trait and implementations

### LINKED ISSUE #229: [M3] Abstraction for transaction signature verification
**Issue summary**

**Other information and links**

### PR DESCRIPTION
**Pull request by willemolding**
**Summary of changes**
Changes introduced in this pull request:
- Introduce `Verifiable` trait in base crate
- Implement trait for `SignedTransaction`, `StagedLedgerDiff`, `ExternalTransition` and others
- Restructure `StagedLedgerDiff` to be more readable and rusty
- Update tests and serialization conversions for updated `StagedLedgerDiff`

**Reference issue to close (if applicable)**

Closes #229 

**Other information and links**

### CODE REVIEW

**Review by hanabi1224 — APPROVED** [review]

File: base/src/staged_ledger_diff.rs [hunk 1]
```diff
@@ -9,46 +9,53 @@
...
-    pub coinbase: CoinBase,
-    pub internal_command_balances: Vec<InternalCommandBalanceData>,
+impl<CTX> Verifiable<CTX> for StagedLedgerPreDiff
+where
+    CTX: Signer<SignedCommandPayload>,
+{
+    // PreDiff is considered valid if:
+    // - all commands are valid
+    fn verify(&self, ctx: &mut CTX) -> bool {
+        self.commands
```
Comment (creativcoder) [inline base/src/staged_ledger_diff.rs:55]: Do we only verify `commands` field for this?

File: base/src/verifiable.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+/// Types that implement Verifiable are able to self-check using intrinsic data (e.g. signatures)
+/// and optionally some context that is required for peforming the verification
+/// e.g. a singleton signature verifier or param required for snark verification
+pub trait Verifiable<CTX> {
+    /// Accepts self and a context and
+    /// returns if the implementor is valid
+    ///
+    /// The context must be a mutable reference to satisfy the requirements of using a
+    /// shared signature checking context which performs self mutation as an optimization
+    fn verify(&self, ctx: &mut CTX) -> bool;
```
Comment (creativcoder) [inline base/src/verifiable.rs:15]: Can the verify operation fail. If yes, probably good to return a `Result` with error context.

Comment (willemolding) [reply base/src/verifiable.rs:15]: I considered this but I think that a failure is just something not being valid. Also returning  a bool keeps it consistent with the proof-systems `verify` function https://github.com/o1-labs/proof-systems/blob/3961f5cfce1f8e2f014ff8afeb50241eb6a161a4/signer/src/schnorr.rs#L71

File: base/src/staged_ledger_diff.rs (hunk 1 above)
Comment (willemolding) [reply base/src/staged_ledger_diff.rs:55]: That and the snark work but I guess that has to wait

File: base/src/verifiable.rs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+//! Traits and helpers for data structures that require validation
+
+/// Types that implement Verifiable are able to self-check using intrinsic data (e.g. signatures)
+/// and optionally some context that is required for peforming the verification
+/// e.g. a singleton signature verifier or param required for snark verification
+pub trait Verifiable<CTX> {
```
[the remark this answers is not in the document]
Comment (willemolding) [reply base/src/verifiable.rs:9]: I'm not sure if I understand what you mean

File: base/src/staged_ledger_diff.rs (hunk 1 above)
Comment (sudo-shashank) [reply base/src/staged_ledger_diff.rs:55]: can we add a todo in this case and mention what's blocking it

**Review by sudo-shashank — APPROVED** [review]

File: base/src/verifiable.rs (hunk 2 above)
Comment (creativcoder) [reply base/src/verifiable.rs:15]: sounds good

File: base/src/verifiable.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (creativcoder) [reply base/src/verifiable.rs:9]: I meant: is this trait just for verification related to proofs, or also for steps that involve validating a block?

**Comment by willemolding** [comment]
To much has changed under this for an easy rebase. Will post a new PR

**Pull request closed without merging**